cloud/aws/actuators/machine/actuator.go

```diff
@@ -183,6 +243,408 @@ func (a *Actuator) Exists(cluster *clusterv1.Cluster, machine *clusterv1.Machine
 	}
 }
 
+// Ensures that the security groups of the machine are correct
+// Returns bool, error
+// Bool indicates if changes were made or not, allowing the caller to decide
+// if the machine should be updated.
+func (a *Actuator) ensureSecurityGroups(machine *clusterv1.Machine, instanceID *string, additionalSecurityGroups []v1alpha1.AWSResourceReference, instanceSecurityGroups map[string]string) (bool, error) {
+	// Get the SecurityGroup annotations
+	annotation, err := a.machineAnnotationJSON(machine, SecurityGroupsLastAppliedAnnotation)
+	if err != nil {
+		return false, err
+	}
+
+	// Now we check to see if we need to mutate any mutable state.
+	// Check if any additions have been made to the instance Security Groups.
+	changed, groupIDs, newAnnotation := a.securityGroupsChanged(
+		annotation,
+		additionalSecurityGroups,
+		instanceSecurityGroups,
+	)
+	if changed {
+		// Finally, update the instance with our new security groups.
+		err := a.ec2.UpdateInstanceSecurityGroups(instanceID, groupIDs)
+		if err != nil {
+			return false, err
+		}
+
+		// And then update the annotation to reflect our new managed state.
+		err = a.updateMachineAnnotationJSON(machine, SecurityGroupsLastAppliedAnnotation, newAnnotation)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	return changed, nil
+}
+
+// Ensure that the tags of the machine are correct
+// Returns bool, error
+// Bool indicates if changes were made or not, allowing the caller to decide
+// if the machine should be updated.
+func (a *Actuator) ensureTags(machine *clusterv1.Machine, instanceID *string, additionalTags map[string]string) (bool, error) {
```

Comment:
Generally I'd like to keep the actuator small and put this kind of logic in some other object or package. Maybe a tags manager or similar